### DISCUSSION TITLE
What happend to tft.readRect() and tft.pushRect()??? Are there bugs?

### DISCUSSION THREAD

**Discussion by davidzhu210** [Bodmer/TFT_eSPI#2440] [Q&A - General]
I am using a 480x320 TFT driven by ST7796.
I pushed a JPEG picture to the TFT using JPEGDecoder.h lib. Then I used tft.readRect() to read a block of pixels to an array. When I used tft.pushRect() to push it back to the screen I saw the block is in a wrong color. The fanny thing is, when I read the block that pushed to the TFT, then push it back to the screen again I can see the correct color. See my codes:

#include <Arduino.h>
#include <TFT_eSPI.h>
#include "drawJpeg.h" // includes SPIFFS.h, JPEGDecoder.h, etc

TFT_eSPI tft = TFT_eSPI();
#define BLOCKPOSX 120
#define BLOCKPOSY 100
#define BLOCKWIDTH 128   
#define BLOCKHEIGHT 128  

uint16_t block[BLOCKWIDTH * BLOCKHEIGHT];  // array to store pixel data in RGB565 format

void setup() {
    Serial.begin(115200);

    tft.init();
    tft.setRotation(3);

    drawJpegMid("/cat.jpg");  // push the picture to the screen, no problem.
    tft.readRect(BLOCKPOSX, BLOCKPOSY, BLOCKWIDTH, BLOCKHEIGHT, pixels565);  // read the pixels to the block

    // push the block back to TFT up-left corner. I see the wrong color!!
    tft.pushRect(0, 0, BLOCKWIDTH, BLOCKHEIGHT, block);

    tft.readRect(0, 0, BLOCKWIDTH, BLOCKHEIGHT, block);  // read the pushed block from the up-left corner
    // push it to another location. I see the correct color!!!
    tft.pushRect(BLOCKPOSX - 120, BLOCKPOSY + 80, BLOCKWIDTH, BLOCKHEIGHT, block);
}

void loop() {}

This is the SPI settings in  "User_Setup.h"
#define SPI_FREQUENCY 40000000
#define SPI_READ_FREQUENCY 20000000*/

I cannot see anything wrong with my code.
Please help.

![IMG_20230222_202957](https://user-images.githubusercontent.com/47712579/220621047-42c5cb12-a42e-4422-a202-6d3fc0dad5de.jpg)

**Comment by Bodmer** [comment]
I pu together a complete example to test and it works fine on my setup, SPI interface ST7796 and RP2040 Pico. Try it:

[TFT_flash_jpg_2440.zip](https://github.com/Bodmer/TFT_eSPI/files/10805452/TFT_flash_jpg_2440.zip)

**Comment by davidzhu210** [comment]
The problem persists. Please see the picture.
I attached "User_Setup" file, just for your convinence.

![IMG_20230223_084223](https://user-images.githubusercontent.com/47712579/220796579-2a6b608c-bd22-4714-9e3f-add402abd762.jpg)
[User_Setup.txt](https://github.com/Bodmer/TFT_eSPI/files/10809419/User_Setup.txt)

**Comment by davidzhu210** [comment]
I added two lines: read the block from (64,64) and put it to another location on TFT and I see the correct color!

void loop() {
    tft.setRotation(0);  // portrait
    tft.fillScreen(TFT_BLACK);

    drawArrayJpeg(EagleEye, sizeof(EagleEye), 0, 0);  // Draw a jpeg image stored in memory at x,y
    delay(2000);
    tft.readRect(0, 0, BLOCKWIDTH, BLOCKHEIGHT, block);
    tft.pushRect(64, 64, BLOCKWIDTH, BLOCKHEIGHT, block);

    delay(2000);
    tft.readRect(64, 64, BLOCKWIDTH, BLOCKHEIGHT, block);     // read the block from (64,64)
    tft.pushRect(64, 224, BLOCKWIDTH, BLOCKHEIGHT, block);  // put it to another location
    delay(2000);
}

![IMG_20230223_085545](https://user-images.githubusercontent.com/47712579/220798961-6ea78a09-9ec7-40a5-877b-83e99dda70ed.jpg)

**Comment by Bodmer** [comment]
Ah! You have a parallel bus interface display.  The SPI speeds are irrelevant and mislead me into testing a SPI display.

I think this is fixed now in the latest master library. You will need to download direct from Github.

**Comment by davidzhu210** [comment]
I downlaoded the latest version 2.5.22
Now I see a different problem, pls. refer to the picture. This is my platformio.ini:

  [env:esp32-s3-devkitc-1]
  platform = espressif32
  board = esp32-s3-devkitc-1
  framework = arduino
  monitor_speed = 115200
  lib_ldf_mode = deep+
  lib_deps = 
	  bodmer/TFT_eSPI@^2.5.22
	  tamctec/TAMC_GT911@^1.0.2
	  fastled/FastLED@^3.5.0
	  bodmer/JPEGDecoder@^1.8.1


![IMG_20230224_082433](https://user-images.githubusercontent.com/47712579/221061895-ff90639b-4f5f-4ea0-ab94-fba26cb27ab7.jpg)

**Comment by davidzhu210** [comment]
Dear Bodmer, I still see the problem (see thea above post). Can you support?

**Reply by Bodmer** [reply]
The display indicates the bytes are swapped in the memory buffer. This is rather odd as I have tested a parallel interface ESP32 S3 with the latest Github library and it works fine with an ST7796 display.

One way to correct this byte swap in your case is to replace this line in your sketch:
```
tft.pushRect(64, 64, BLOCKWIDTH, BLOCKHEIGHT, block);
```
with these lines:
```
  bool swap = tft.getSwapBytes();
  tft.setSwapBytes(true);  // If still wrong image, try false
  tft.pushImage(64, 64, BLOCKWIDTH, BLOCKHEIGHT, block);
  tft.setSwapBytes(swap);
```

**Reply by Bodmer** [reply]
I have raised the github master version to 2.5.23 so you can make sure you have the latest update.

Since my setup with the same hardware works fine, other posibilities are:
1. the setup file is not being read correctly and the wrong display type is enabled (run Read_User_Setup sketch to verify)
2. the display you have is not an ST7796 type

**Comment by davidzhu210** [comment]
These are my etst results:
1. I run "Read_User_Setup" sketch and confirmed the driver is 7796. The following are the readings:
[code]
TFT_eSPI ver = 2.5.22
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = Parallel
Display driver = 7796
Display width  = 320
Display height = 480

TFT_CS   = GPIO 10
TFT_DC   = GPIO 9
TFT_RST  = GPIO 14
TFT_WR   = GPIO 15
TFT_RD   = GPIO 6

TFT_D0   = GPIO 1
TFT_D1   = GPIO 2
TFT_D2   = GPIO 7
TFT_D3   = GPIO 8
TFT_D4   = GPIO 3
TFT_D5   = GPIO 18
TFT_D6   = GPIO 17
TFT_D7   = GPIO 16

TFT_BL           = GPIO 4
TFT_BACKLIGHT_ON = LOW

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

[/code]


2.  In Arduino lib I cannot find ver. 2.5.23. So I still use 2.5.22 for my test.
3. I modified the sketch:
![image](https://user-images.githubusercontent.com/47712579/222023379-57826e72-fc5f-40c6-8474-fe89da220ad1.png)
The problem persists, see below:
![IMG_20230301_094803](https://user-images.githubusercontent.com/47712579/222023883-56a2d11b-24a4-4239-83c8-cedb94a32995.jpg)
4. I changed tft.setSwapBytes(true) to tft.setSwapBytes(false). the problem is still there:
![IMG_20230301_095037](https://user-images.githubusercontent.com/47712579/222024183-b6a76a43-f4d0-408b-b1fe-d55862e101a8.jpg)

**Reply by Bodmer** [reply]
Download the library direct from Github and install. V2.5.23 has not been released yet.
